src/node_url_pattern.cc

```diff
@@ -0,0 +1,571 @@
+      init.hostname = std::string(utf8_value);
+    } else if (component == env->port_string()) {
+      init.port = std::string(utf8_value);
+    } else if (component == env->pathname_string()) {
+      init.pathname = std::string(utf8_value);
+    } else if (component == env->search_string()) {
+      init.search = std::string(utf8_value);
+    } else if (component == env->hash_string()) {
+      init.hash = std::string(utf8_value);
+    } else if (component == env->base_url_string()) {
+      init.base_url = std::string(utf8_value);
+    }
+  };
+  for (const auto& component : components) {
+    if (obj->Get(env->context(), component).ToLocal(&value)) {
+      if (value->IsString()) {
+        Utf8Value utf8_value(isolate, value);
+        set_parameter(component, utf8_value.ToStringView());
+      }
+    }
+  }
+  return init;
+}
+
+Local<Object> URLPattern::URLPatternComponentResult::ToJSObject(
+    Environment* env, const ada::url_pattern_component_result& result) {
+  auto isolate = env->isolate();
+  Local<Name> names[] = {env->input_string(), env->groups_string()};
+  Local<Value> values[] = {
+      ToV8Value(env->context(), result.input).ToLocalChecked(),
+      ToV8Value(env->context(), result.groups).ToLocalChecked(),
+  };
+  return Object::New(isolate, Null(isolate), names, values, 2);
+}
+
+Local<Object> URLPattern::URLPatternResult::ToJSObject(
+    Environment* env, const ada::url_pattern_result& result) {
+  auto isolate = env->isolate();
+  Local<Name> names[]{
+      env->protocol_string(),
+      env->username_string(),
+      env->password_string(),
+      env->hostname_string(),
+      env->port_string(),
+      env->pathname_string(),
+      env->search_string(),
+      env->hash_string(),
+  };
+  Local<Value> values[] = {
+      URLPatternComponentResult::ToJSObject(env, result.protocol),
+      URLPatternComponentResult::ToJSObject(env, result.username),
+      URLPatternComponentResult::ToJSObject(env, result.password),
+      URLPatternComponentResult::ToJSObject(env, result.hostname),
+      URLPatternComponentResult::ToJSObject(env, result.port),
+      URLPatternComponentResult::ToJSObject(env, result.pathname),
+      URLPatternComponentResult::ToJSObject(env, result.search),
+      URLPatternComponentResult::ToJSObject(env, result.hash),
+  };
+  DCHECK(arraysize(names) == 8);
+  DCHECK(arraysize(values) == 8);
+  return Object::New(isolate, Null(isolate), names, values, 8);
+}
+
+ada::url_pattern_options URLPattern::URLPatternOptions::FromJsObject(
+    Environment* env, Local<Object> obj) {
+  ada::url_pattern_options options{};
+  Local<Value> ignore_case;
+  if (obj->Get(env->context(),
+               FIXED_ONE_BYTE_STRING(env->isolate(), "ignoreCase"))
+          .ToLocal(&ignore_case)) {
+    options.ignore_case = ignore_case->IsTrue();
+  }
+  return options;
+}
+
+MaybeLocal<Value> URLPattern::Hash() const {
+  auto context = env()->context();
+  return ToV8Value(context, url_pattern_.get_hash());
```

Comment:
I think the key challenge here is that this will copy the string on every call. Any chance of memoizing the string once created.

Reply:
I'll follow up on a different PR for this